### PR TITLE
[doc] [dev] Document memtrace use

### LINKED ISSUE #13161: [performance] memtrace explorations
I am opening this issue to share / discuss about evaluation of Coq with `memtrace`; there is an excellent post explaining how to use it:

https://blog.janestreet.com/finding-memory-leaks-with-memtrace/

I have done a few experiments myself and the picture looks a bit discouraging, for example w.r.t. allocation in proofs due to the tactic engine, and some other strange spots that I cannot explain. This is likely known by some / most of the developers already, but maybe we could use the graphical tool to better identify some critical spots in the current implementation.

I was going to post some findings, but I'm afraid I don't understand the data well yet so I will refraind until I can gain a better understanding.

Patch to try in Coq master:
```patch
diff --git a/topbin/coqc_bin.ml b/topbin/coqc_bin.ml
index 2006aa2d0f..adea4ee55f 100644
--- a/topbin/coqc_bin.ml
+++ b/topbin/coqc_bin.ml
@@ -10,4 +10,5 @@
 
 (* Main coqc initialization *)
 let _ =
+  Memtrace.trace_if_requested ();
   Coqc.main ()
diff --git a/topbin/dune b/topbin/dune
index 46052c81e5..fcde1769c4 100644
--- a/topbin/dune
+++ b/topbin/dune
@@ -25,7 +25,7 @@
  (public_name coqc)
  (package coq)
  (modules coqc_bin)
- (libraries coq.toplevel)
+ (libraries coq.toplevel memtrace)
  (modes native byte)
  ; Adding -ccopt -flto to links options could be interesting, however,
  ; it doesn't work on Windows
```
Then, for example, if you want to try to compile file `Byte.v`, first make a copy to `MyByte.v` [that's to avoid issues with actifacts]. Then:
```
$ dune build theories/Strings/Byte.vo
$ MEMTRACE=trace-byte.tcr _build/install/default/bin/coqc MyByte.v
$ memtrace-viewer trace-byte.tcr
```

### PR DESCRIPTION
**Pull request by ejgallego**
Closes #13161

Should we maybe clean up the other memory profiling doc?

It seems stale (cc @ppedrot )

### CODE REVIEW

File: dev/doc/profiling.md [hunk 1]
```diff
@@ -70,21 +76,42 @@ perf script --time '0%-100%'  |
...
+To profile a file, we recommend you make a copy (to avoid issues with
+artifacts), then you can do:
+
+```
+dune build theories/Strings/Byte.vo  # to build deps of Byte
+MEMTRACE=trace-byte.tcr dune exec -- dev/shim/coqc MyByte.v
+memtrace-viewer trace-byte.tcr
+```
+
+## Memory (legacy)
```
Comment (ppedrot) [inline dev/doc/profiling.md:98]: I agree that this section should go away.

File: dev/doc/profiling.md [hunk 2]
```diff
@@ -70,21 +76,42 @@ perf script --time '0%-100%'  |
...
         ) |
         flamegraph.pl
-~~~~~
+```
 
-## Memory
+## Memory (memtrace)
+
+[memtrace](https://github.com/janestreet/memtrace) is a client library
+for OCaml's Memprof statistical memory profiler.
```
Comment (ppedrot) [inline dev/doc/profiling.md:84]: You should mention somewhere that for this to work in Coq you must have memtrace installed when compiling Coq.

Comment (ejgallego) [reply dev/doc/profiling.md:84]: Done thanks!

File: dev/doc/profiling.md (hunk 1 above)
Comment (ejgallego) [reply dev/doc/profiling.md]: Done thanks, let's see how it looks now.

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
@coqbot merge now

**Pull request merged**